### PR TITLE
Ruby agent: gRPC instrumentation

### PR DESCRIPTION
**Pull request by fallwith**
gRPC instrumentation documentation for New Relic Ruby agent v8.10.0+

### CODE REVIEW

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+INFO : Installing New Relic supported gRPC_Client instrumentation using Prepend
+INFO : Installing New Relic supported gRPC_Server instrumentation using Prepend
+```
+
+
+## Usage
+
+As long as the instrumentation is working properly (see "Verification" above to confirm this), all instrumentation performed by the New Relic Ruby agent should be fully automatic and not require any modifications whatsoever to any of your existing gRPC client and/or server code.
+
+The New Relic Ruby agent's gRPC instrumentation can be used for New Relic Distributed Tracing in conjunction with any other New Relic agent that supports gRPC. gRPC requests can be sent from a Ruby-based client to a server based on another language, and a Ruby-based server can field requests coming from a client based on another language.
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx:72]:
```suggestion
The New Relic Ruby agent's gRPC instrumentation includes distributed tracing headers in spans, which allows gRPC client and server calls to be connected to any other service that supports W3C distributed tracing. gRPC requests can be sent from a Ruby-based client to a server based on another language, and a Ruby-based server can field requests coming from a client based on another language.
```

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+<Callout variant="important">
+  gRPC client requests will not be reported to New Relic unless they are spawned when an existing New Relic transaction is present. Typically this means that a Rails controller method or background job process is active and responsible for the invocation of the gRPC client request. The request will then roll up under that outer wrapper's transaction in the New Relic web UI.
+
+  gRPC server request handlings will always generate a new New Relic transaction and will not require nor roll-up to any pre-existing New Relic transaction.
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx:77]:
```suggestion
  gRPC server requests will always generate a new New Relic transaction and will not require nor roll-up to any pre-existing New Relic transaction.
```

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+### Client Instances
+
+On the [Summary page](/docs/apm/applications-menu/monitoring/apm-overview-page), gRPC client requests will appear as Web transactions and contribute data to the "Web transactions time", "Throughput", and "Error rate" charts.
+
+On the [Distributed tracing page](/docs/apm/agents/ruby-agent/configuration/distributed-tracing-ruby-agent/), select (click on) a trace group, then click on the trace group name again, then click on "Expand all" and then select (click on) a specific nested client trace. With a specific client trace selected, select the Attributes tab on the right. The `grpc_status`, `grpc_type`, `http.method`, and `http.url` attributes should all appear. In the event of a failed client request (which will have a `grpc_status` value other than `0`), an additional `grpc_message` attribute will appear with the gRPC status message.
+
+If the gRPC client requests are being sent to a gRPC server with New Relic gRPC instrumentation (based on Ruby or another language), then the Distributed tracing trace map should display both the client and server applications and report on how many distinct call types were performed between them.
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx:91]:
```suggestion
If the gRPC client requests are being sent to a gRPC server with New Relic gRPC instrumentation (based on Ruby or another language), then the distributed tracing trace map should display both the client and server applications and report on how many distinct call types were performed between them.
```

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+In the event of a gRPC-based Ruby exception, information about the exception will appear on the Events > Errors page.
+
+### Server Instances
+
+On the [Summary page](/docs/apm/applications-menu/monitoring/apm-overview-page), gRPC client requests will appear as Web transactions and contribute data to the "Web transactions time", "Throughput", and "Error rate" charts.
+
+On the [Distributed tracing page](/docs/apm/agents/ruby-agent/configuration/distributed-tracing-ruby-agent/), select (click on) a client trace group, and then click on the trace group name again. Then from the resulting page's list, click on an individual gRPC server trace and select (click on) the Attributes tab on the right. The `request.grpc_type`, `request.method`, and `request.uri` attributes should all appear.
+
+If the gRPC server handled requests originated from a gRPC client with New Relic gRPC instrumentation (based on Ruby or another language), then the Distributed tracing trace map should display both the client and server applications and report on how many distinct call types were performed between them.
```
Comment (kaylareopelle) [inline src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx:101]:
```suggestion
If the gRPC server handled requests originated from a gRPC client with New Relic gRPC instrumentation (based on Ruby or another language), then the distributed tracing trace map should display both the client and server applications and report on how many distinct call types were performed between them.
```

**Review by kaylareopelle — COMMENTED** [review]
In my opinion, this document should be considered as our future standard for new instrumentation! Great work!

**Comment by kaylareopelle** [comment]
I wonder if it might be helpful to customers to have a table with the attributes for client/server objects, kind of like we have in the implementation guide.

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx [hunk 5]
```diff
@@ -0,0 +1,119 @@
+---
+title: gRPC instrumentation
+tags:
+  - Agents
+  - Ruby agent
+  - Instrumented gems
```
[the remark this answers is not in the document]
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx:6]: None of the other .mdx files in the directory have anything beyond this list of 3, so I wanted the new .mdx file to match.

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+metaDescription: The New Relic Ruby agent automatically instruments gRPC clients and servers.
+---
+
+## Overview
+
+The [New Relic Ruby agent](/docs/agents/ruby-agent/getting-started/new-relic-ruby) automatically instruments [gRPC](https://grpc.io/) clients and servers that use the [grpc](https://rubygems.org/gems/grpc) gem (gem version 1.0.0 or higher). All communication types - unary, bidirectional, client-streaming, and server-streaming - are supported.
```
[the remark this answers is not in the document]
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx]: committed, thank you!

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+<Callout variant="important">
+  gRPC instrumentation requires Ruby agent version 8.10.0 or higher.
+</Callout>
+
+
+## Configuration (All settings are optional)
+
+The instrumentation of gRPC clients and servers can be configured separately. By default, both instrumentations are set to `auto`, which is short for "automatic". The agent's automatic behavior for each is to enable the instrumentation by attempting to leverage Ruby module prepend behavior at first and falling back on Ruby method chaining behavior if known compatibility issues are encountered. Optionally, the instrumentation setting can be set to a value of `disabled` to disable the instrumentation, `prepend` for forced module prepend behavior, or `chain` for forced method chaining behavior.
```
[the remark this answers is not in the document]
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx]: suggestion committed, thank you

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx (hunk 1 above)
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx]: committed, thanks

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx (hunk 2 above)
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx:77]: "handlings" was a deliberate term to incorporate streaming (one long running open connection with multiple payloads across what from an HTTP perspective might be a single "request"), but this approach doesn't require such a lengthy footnote, merging.

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx (hunk 3 above)
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx]: committed, thank you

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx (hunk 4 above)
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx]: committed, thank you

File: src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+  **gRPC client**
+  - config file parameter: `instrumentation.grpc_client`
+  - environment variable: `NEW_RELIC_INSTRUMENTATION_GRPC_CLIENT`
+
+  **gRPC server**
+  - config file parameter: `instrumentation.grpc_server`
+  - environment variable: `NEW_RELIC_INSTRUMENTATION_GRPC_SERVER`
+
+  Examples:
```
[the remark this answers is not in the document]
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/instrumented-gems/grpc-instrumentation.mdx]: committed 8885ad3

**Comment by fallwith** [comment]
> I wonder if it might be helpful to customers to have a table with the attributes for client/server objects, kind of like we have in the implementation guide.

I'd like to hold off on doing so for now. I'm unaware of that being done for any other instrumentation, so it'd require some best practices conversations to come up with a standard if this was the first doc portraying that content. And for gRPC specifically, I would prefer that that type of thing be left up to the doc writers whenever it comes time to standardize on gRPC support across all agents with how things are consistently documented, as we've done recently for CLM and logging. It'd be really great to have better gRPC instrumentation documentation available to compare with from the other agents.

**Review by rhetoric101 — COMMENTED** [review]
@fallwith this is great work! Thanks for adding all of this new information!

I decided to make some direct changes to your docs instead of making suggestions here. Please take a look at the changes to confirm that I didn't change the meaning, especially in the main gRPC doc. Feel free to make corrections as you see fit.

File: src/content/docs/release-notes/agent-release-notes/ruby-release-notes/ruby-agent-8100.mdx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    The following new configuration parameters have been added for gRPC. All are optional.
+
+    | Configuration name | Default | Behavior |
+    | ----------- | ----------- |----------- |
+    | `instrumentation.grpc_client` | auto | Set to 'disabled' to disable, set to 'chain' if there are module prepending conflicts |
+    | `instrumentation.grpc_server` | auto | Set to 'disabled' to disable, set to 'chain' if there are module prepending conflicts |
+    | `instrumentation.grpc.host_denylist` | "" |  Provide a comma delimited list of host regex patterns (ex: "private.com$,exception.*") |
```
[the remark this answers is not in the document]
Comment (fallwith) [reply src/content/docs/release-notes/agent-release-notes/ruby-release-notes/ruby-agent-8100.mdx:21]: Good question. The config file holds a string, and the quotes are optional unless there's a space involved, given that it's in YAML format. Then at startup, the agent will parse the string and convert the delimited values into an array. So both a string and an array are appropriate here.

Looking through our [configuration documentation](https://docs.newrelic.com/docs/apm/agents/ruby-agent/configuration/ruby-agent-configuration/#labels), I searched for "delimited". Every value there appears appears to use a double quoted string for this type of thing.

File: src/content/docs/apm/agents/ruby-agent/features/ruby-codestream-integration.mdx [hunk 10]
```diff
@@ -14,7 +14,7 @@ Display context-sensitive APM data directly in your IDE by integrating [New Reli
 First, [install](/docs/codestream/start-here/install-codestream) the New Relic CodeStream extension into your supported IDE of choice and log in.
 
 <Callout variant="important">
-  The New Relic CodeStream integration is available in Ruby agent version 8.8.0 and higher and is disabled by default. To enable the integration, either set `code_level_metrics.enabled: true` in `newrelic.yml` or `NEW_RELIC_CODE_LEVEL_METRICS_ENABLED=true` as an environment variable.
+  The New Relic CodeStream integration is available in Ruby agent version 8.8.0 and higher. For versions 8.10.0 and higher, the integration is enabled by default. Versions 8.8.0 and 8.9.0 must have the integration enabled via configuration. To enable the integration, either set `code_level_metrics.enabled: true` in `newrelic.yml` or `NEW_RELIC_CODE_LEVEL_METRICS_ENABLED=true` as an environment variable.
```
[the remark this answers is not in the document]
Comment (fallwith) [reply src/content/docs/apm/agents/ruby-agent/features/ruby-codestream-integration.mdx:17]: Good call. I have committed ac2235d1834d298797f066aeeb2d0aa49bfe78ab to update that file. Please feel free to make any changes you wish to it.

**Comment by fallwith** [comment]
> I decided to make some direct changes to your docs instead of making suggestions here. Please take a look at the changes to confirm that I didn't change the meaning, especially in the main gRPC doc. Feel free to make corrections as you see fit.

@rhetoric101 your changes look fantastic, thanks very much! The formatting, layout, linking, Gatsby fixes, etc. changes are all great, and your rephrasing is much appreciated. The original text was cobbled together from multiple iterations made by multiple authors working across multiple developer facing documents, and it showed. You've made things more coherent, approachable, and singular in voice.

**Review by rhetoric101 — APPROVED** [review]
@fallwith thanks for working through these suggestions. I think we're all set now!

**Pull request merged**